README.md
# OpenBCI WiFi Shield Library

<p align="center">
  <img alt="banner" src="/images/WiFi_front_product.jpg/" width="400">
</p>
<p align="center" href="">
  Make programming with OpenBCI reliable, easy, research grade and fun!
</p>

## Welcome!

First and foremost, Welcome! :tada: Willkommen! :confetti_ball: Bienvenue! :balloon::balloon::balloon:

Thank you for visiting the OpenBCI WiFi Shield Library repository.

This document (the README file) is a hub to give you some information about the project. Jump straight to one of the sections below, or just scroll down to find out more.

* [What are we doing? (And why?)](#what-are-we-doing)
* [Who are we?](#who-are-we)
* [What do we need?](#what-do-we-need)
* [How can you get involved?](#get-involved)
* [Get in touch](#contact-us)
* [Find out more](#find-out-more)
* [Installation](#install)
* [Building](#build)
* [Running](#running)
* [License](#license)

## What are we doing?

### The problem

* Users continuously struggle to get prerequisites properly installed to get current OpenBCI Cyton and Ganglion, hours/days/weeks are wasted just _trying to get the data_.
* Bluetooth requires you to stay close to your computer, if you go to far, data is lost and the experiment is over.
* Bluetooth is too slow for transmitting research grade EEG, researchers want 1000Hz (samples per second), bluetooth with 8 channels is limited to 250Hz and with 16 channels limited to 125Hz.
* Bluetooth is unreliable when many other Bluetooth devices are around, demo device or use in busy real life experiment is not reliable. (think grand central station at rush hour)
* Bluetooth requires data to be sent to desktops in raw or compressed form, must use other node modules to parse complex byte streams, prevents from running in browser.
* OpenBCI Cyton (8 and 16 channel) with Bluetooth cannot go to any mobile device because of required Bluetooth-to-USB "Dongle". Must use USB port on Desktop/Laptop computers.
* Bluetooth on Ganglion requires low level drivers to use computers bluetooth hardware.
* The OpenBCI Cyton and Ganglion must transmit data to another computer over Bluetooth before going to the cloud for storage or analytics
* OpenBCI Cyton Dongle FTDI virtual comm port drivers have high latency by default which limits the rate at which new data is made available to your application to twice a second when it should get data as close to 250 times a second.

So, if even the very best developers integrate the current easy to use Cyton and Ganglion, they are still burdened by the limitations of the physical hardware on the OpenBCI system.

### The solution

The OpenBCI WiFi Shield Library will:

* Find, connect, sync, and configure (e.g. set sample rate) with carrier boards (Ganglion or Cyton or Cyton with Daisy) automatically
* Use TCP over WiFi to prevent packet loss
* When using UDP, offer a way to use send packets redundantly
* Enable streaming of high speed (samples rates over 1000Hz), low latency (by default, send data every 10ms), research grade EEG (no lost data) directly to any internet connected device (i.e. iPhone, Android, macOS, Windows, Linux, Raspberry Pi 3)
* Use WiFi direct to create a stable wireless transmission system even in crowded areas

Using WiFi physically solves limitations with the current state-of-the-art open source bio sensor. The goal for the WiFi Shield firmware was to create a [_one up_](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&cad=rja&uact=8&ved=0ahUKEwjF7pax67PWAhUH6oMKHdfJAcgQFggoMAA&url=https%3A%2F%2Fwww.electroimpact.com%2FWhitePapers%2F2008-01-2297.pdf&usg=AFQjCNHSyVXxRNtkFrmPiRqM5WqHWdO9-g) data pipeline, where scientific data in JSON is sent instead of raw/compressed ADC counts (yuk!) to ***make programming with OpenBCI reliable, easy, research grade and fun!***

## Who are we?

The founder of the OpenBCI WiFi Shield Library is [AJ Keller][link_aj_keller] with [Push The World][link_ptw]. As we continue to stabilize the hardware, the people contributing to this repo is the [OpenBCI][link_openbci] at large. WiFi library contributors want to stream data _fast_ from their bio-sensors.

<a href="https://www.mozillascience.org/about">
  <img
    src="http://mozillascience.github.io/working-open-workshop/assets/images/science-fox.svg"
    align="right"
    width=140
  </img>
</a>

[AJ][link_aj_keller] is an invited member of the 4th cohort [Open Leaders Cohort][link_openleaderscohort] of the [Mozilla Science Lab][link_mozsci] who brought together open science advocates from around the world to participate in the first [Working Open Workshop][link_mozwow] in Berlin in February 2016. The [training exercises][link_mozwow] (which are free and easy to reuse) focused on how to build and effectively engage communities so they can work together to develop tools and resources for the greater good.

## What do we need?

**You**! In whatever way you can help.

We need expertise in programming, user experience, software sustainability, documentation and technical writing and project management.

We'd love your feedback along the way.

Our primary goal is to make programming with OpenBCI reliable, easy, research grade and fun, and we're excited to support the professional development of any and all of our contributors. If you're looking to learn to code, try out working collaboratively, or translate you skills to the digital domain, we're here to help.

## Get involved

If you think you can help in any of the areas listed above (and we bet you can) or in any of the many areas that we haven't yet thought of (and here we're *sure* you can) then please check out our [contributors' guidelines](CONTRIBUTING.md) and our [roadmap](ROADMAP.md).

Please note that it's very important to us that we maintain a positive and supportive environment for everyone who wants to participate. When you join us we ask that you follow our [code of conduct](CODE_OF_CONDUCT.md) in all interactions both on and offline.


## Contact us

If you want to report a problem or suggest an enhancement we'd love for you to [open an issue](../../issues) at this github repository because then we can get right on it. But you can also contact [AJ][link_aj_keller] by email (pushtheworldllc AT gmail DOT com) or on [twitter](https://twitter.com/aj-ptw).

You can also hang out, ask questions and share stories in the [OpenBCI NodeJS room](https://gitter.im/OpenBCI/OpenBCI_NodeJS?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge) on Gitter.

## Find out more

You might be interested in:

* Purchase a [WiFi Shield from OpenBCI](https://shop.openbci.com/collections/frontpage/products/wifi-shield?variant=44534009550)
* An example of the WiFi Shield: [DefaultWifiShield][link_wifi_default]

And of course, you'll want to know our:

* [Contributors' guidelines](CONTRIBUTING.md)
* [Roadmap](ROADMAP.md)

## Thank you

Thank you so much (Danke schön! Merci beaucoup!) for visiting the project and we do hope that you'll join us on this amazing journey to make programming with OpenBCI fun and easy.

## <a name="install"></a> Installation:

Use the `DefaultWifiShield.ino` in examples.

You need to clone the [ESP8266 library from github](https://github.com/esp8266/Arduino#using-git-version).

For the firmware that runs on ESP8266 see [OpenBCI_Wifi](https://github.com/PushTheWorld/OpenBCI_Wifi) on PushTheWorld’s github repository.

For the software that runs on OpenBCI, we will start from scratch with an ino file that just does SPI.

The ESP8266 runs arduino as well so there are several installs that need to happen in order to make that work; repo: [ESP8266 Arduino](https://github.com/esp8266/Arduino) and follow the instructions for [Using Git Version](https://github.com/esp8266/Arduino/blob/master/README.md#using-git-version) to install properly. We have to do this because we need the “SPISlave” feature that has not been published in their latest release.

The library we are interested in using is the [`SPISlave`](https://github.com/esp8266/Arduino/tree/master/libraries/SPISlave).


## <a name="build"></a> Building and Flashing:

There are two steps to flashing code. Compiling the codebase, and then flashing it onto the wifi module.
Makefiles make it a lot easier. Especially if you use [makeEspArduino](https://github.com/plerup/makeEspArduino). Follow the instructions from the project's github.

The `make` command runs the default commands from `makeEspArduino` with OpenBCI_Wifi specific variables provided in the `Makefile`. By default builds the default sketch in the `examples/` directory.

-   To build *and* flash run:

    `make flash`

-   To build a different sketch, provide the (full or relative) path to the sketch file, e.g.:

    `make SKETCH=examples/WifiShieldJSON/WifiShieldJSON.ino`

-   To build *and* flash with a specific sketch file:

    `make SKETCH=examples/WifiShieldJSON/WifiShieldJSON.ino flash`

-   If your board is *not using* port `/dev/ttyUSB0`, then pass the `UPLOAD_PORT` variable, e.g.:

    `make UPLOAD_PORT=/dev/cu.usbserial-AI02BAAJ flash`

-   To build *and* flash if your board is *not using* port `/dev/ttyUSB0`, then pass the `UPLOAD_PORT` variable, e.g.:

    `make SKETCH=examples/WifiShieldJSON/WifiShieldJSON.ino UPLOAD_PORT=/dev/cu.usbserial-AI02BAAJ flash`

## <a name="running"></a> Running:

There are two modes that the OpenBCI WiFi shield can run in

<p align="center">
  <img alt="soft-ap" src="/images/wifi_flow_softap.png/" width="500">
</p>
<p align="center" href="">
  WiFi Direct Mode (WiFi<->Computer)
</p>
<p align="center">
  <img alt="soft-ap" src="/images/wifi_flow_station.png/" width="500">
</p>
<p align="center" href="">
  WiFi Station Mode (WiFi<->Router<->Any Internet Device)
</p>

### <a name="license"></a> License:

MIT

[link_aj_keller]: https://github.com/aj-ptw
[link_shop_wifi_shield]: https://shop.openbci.com/collections/frontpage/products/wifi-shield?variant=44534009550
[link_shop_ganglion]: https://shop.openbci.com/collections/frontpage/products/pre-order-ganglion-board
[link_shop_cyton]: https://shop.openbci.com/collections/frontpage/products/cyton-biosensing-board-8-channel
[link_shop_cyton_daisy]: https://shop.openbci.com/collections/frontpage/products/cyton-daisy-biosensing-boards-16-channel
[link_ptw]: https://www.pushtheworldllc.com
[link_openbci]: http://www.openbci.com
[link_mozwow]: http://mozillascience.github.io/working-open-workshop/index.html
[link_wifi_default]: examples/DefaultWifiShield/DefaultWifiShield.ino
[link_openleaderscohort]: https://medium.com/@MozOpenLeaders
[link_mozsci]: https://science.mozilla.org


changelog.md
# v2.0.5

### Bug Fixes

* Setting up SPI slave too early could lead to crash

# v2.0.4

### Bug Fixes

* Access point name not set until wifi manager was ran for the first time.

# v2.0.3

### Enhancements

* Reduce buffer size for better wifi manager
* Fix LED flash when there are stored creds and shield attempts to connect
* Add update wifi link to home page

### Work In Progress

* Hitting /wifi on some devices did not trigger wifi manager

### Third party

* Now building with v2.4.0 from ESP8266/Arduino

# v2.0.2

* Reduce buffer size for better OTA Updates
* Add firmware version to homepage

# v2.0.1

* Updates the swaggerhub to 2.0.0 to match new routes.
* Removed ability to change output on UDP to JSON because not supporting JSON right now.

# v2.0.0

### New Features

* UDP support
* WiFi Direct default support (thanks @jnaulty)
* New docs (code of conduct, roadmap, contributing, readme)
* When in WiFi direct mode, connect PC to WiFi Shield hotspot, and navigate to `192.168.4.1`
* Now visit `192.168.4.1/wifi` in your browser to connect wifi shield to a new network
* To erase your network credentials and use wifi direct, use GUI or go to ip address of wifi shield and go to /wifi/delete

### Breaking Changes

* No longer doing captive portal by default, must join network and go to 192.168.4.1

### Work In Progress

* MQTT Secure
* Stability of wifi shield #48

## Beta 2

### Bug Fixes

* Switch to Station if connect to wifi network success.

## Beta 1

### Bug Fixes

* Can't connect to shield #51

# v1.3.0

At this time, can no longer support both RAW and JSON modes with a single binary file. Need optimized builds. We ended up having two modes that didnt work that well instead of highly optimized builds for the different operating modes.

### Breaking Changes

* For RAW only (can't do Raw to JSON or MQTT) - upload the binary for `DefaultWifiShield_v1.3.0.bin`
  * This is the default firmware to be shipped with WiFi Shields
  * Sample rates of 1000Hz with Cyton (with or without Daisy) possible
  * Use the OpenBCI_GUI/OpenBCIHub
* For JSON over (TCP/MQTT) (can't use OpenBCI GUI) - upload the binary for `WifiShield_RawToJSON_v1.3.0.bin`
  * Only tested with 200Hz with Ganglion
  * Only tested with 250Hz with Cyton

# v1.2.0

### Bug Fixes

* Allow for CORS on HTTP server for every call. Closes #41
* Extend max tcp packets per send
* Optimize JSON packet streaming for better performance, allows for latency up to 39ms for 8 channel and 4 channel and 25ms for 16 channel when streaming out JSON
* Changed return of route of /cloud

# v1.1.4

### Bug Fixes

* Calling to /tcp with POST could result in a 502 error when no characters found in _command_ key and also when there was a timeout waiting for the attached board to respond. Changed error when no characters found in _command_ key to 505.

# v1.1.3

### Enhancements

* Spurious code clean up

### Bug Fixes

* Fixed potential issue with gains
* Fixed issue where `outputString` was not cleared after timeout.

# v1.1.2

### Bug Fixes

* Fix bug where DELETE /wifi was not working on some modules.

# v1.1.1

### Bug Fixes

* Extend number of packets per TCP send for Cyton with Daisy board.

# v1.1.0

### Bug Fixes

* Extend raw ring buffer by 1900 packets to 2000. Really helped with dropped packets.

# v1.0.1

### Bug Fixes

* Extended timeout for command passthrough to allow for long messages such as '?' on cyton to be passed through in one message.
* Extended the length of raw buffer to allow for more packets to be sent when at high sample rate.

# v1.0.0

Add /cloud route for cloudbrain redirect.

Lots of stability. Added 596 tests to ensure the core wifi engine core works as expected every time. This is very close to what will be released with the initial production run.

# v0.2.1

### Bug Fixes

* Timestamp was not correct due to offset issue.

# v0.2.0

### Bug Fixes

* When mac address was less than 16 i.e. 0x0X where X is 0-15 the device name would be like "OpenBCI-CF1" instead of "OpenBCI-0CF1".

### Enhancements

* Over the air uploading working! Woo yess go to `/update` and boom! Victory!

# v0.1.2

### Enhancements

* Removed WifiClientPrint dependency from make file.

### New Features

* Add route `/board` to get board type, connection status, number of channels and gains for each channel

# v0.1.1

### Bug Fixes

* Timestamps in JSON are the right unit now. They are 64bit integers.
* Using root.printTo(clientTCP) takes like half a second, switched to print to string and that seemed to speed everything up.
* Getting close to having the JSON working better.

### Known Issues

* Timestamps fall out of sync:
```bash
{"chunk":[{"timestamp":1498731994785376,"data":[3718123.25,3720872.5,3718749.25,3720179.75,3718101,3715262.25,Infinity,3724404]},{"timestamp":1498731994789373,"data":[3718078.5,0,1.942749,3717989.25,3720872.5,3718659.75,3720246.75,3718056.25]},{"timestamp":1498731994793373,"data":[3717989.25,3720872.5,3718659.75,3720246.75,3718056.25,3715262.25,Infinity,3724560.5]}],"count":17488}
{"chunk":[{"timestamp":1498731994797365,"data":[3717855,3721051.25,3718570.25,3720179.75,3718391.5,3715329.25,Infinity,3724337]},{"timestamp":1498731994801403,"data":[Infinity,3724605.25,0,1.942734,1.942711,4.014413,4.050549,0]},{"timestamp":1498731994805362,"data":[3717899.75,3720760.75,3718548,3720492.5,3718190.25,3715217.5,Infinity,3724605.25]}],"count":17489}
{"chunk":[{"timestamp":1498731994809357,"data":[3718011.5,3720895,3718570.25,3720202,3718391.5,3715351.75,Infinity,3724426.5]},{"timestamp":1498731994813353,"data":[1.954601,8.110682,0,3720312,0,0,1.919922,0]},{"timestamp":1498731994817347,"data":[3717966.75,3720515,3718324.5,3720246.75,3718235,3715485.75,Infinity,3724448.75]},{"timestamp":1498731994821341,"data":[3717832.75,3720805.5,3718525.5,3720068,3718011.5,3715374,Infinity,3724538.25]}],"count":17490}

OpenBCI_Wifi_Definitions.h
/**
* Name: OpenBCI_Wifi_Definitions.h
* Date: 8/30/2016
* Purpose: This is the header file for the OpenBCI wifi definitions.
*
* Author: Push The World LLC (AJ Keller)
*/

#ifndef __OpenBCI_Wifi_Definitions__
#define __OpenBCI_Wifi_Definitions__

#define SOFTWARE_VERSION "v2.0.0.0"

#define ADS1299_VREF 4.5
#define MCP3912_VREF 1.2
#define ADC_24BIT_RES 8388607.0
#define ADC_24BIT_MAX_VAL_NANO_VOLT 8388607000000000.0
#define BYTES_PER_SPI_PACKET 32
#define BYTES_PER_OBCI_PACKET 33
#define DEBUG 1
#define MAX_SRV_CLIENTS 2
#define BYTES_PER_CHANNEL 3
#ifdef RAW_TO_JSON
  #define NUM_PACKETS_IN_RING_BUFFER_JSON 28
  #define NUM_PACKETS_IN_RING_BUFFER_RAW 1
#else
  #define NUM_PACKETS_IN_RING_BUFFER_RAW 200
  #define MAX_PACKETS_PER_SEND_TCP 42
#endif
#define WIFI_SPI_MSG_LAST 0x01
#define WIFI_SPI_MSG_MULTI 0x02
#define WIFI_SPI_MSG_GAINS 0x03
#define MAX_CHANNELS 16
#define MAX_CHANNELS_PER_PACKET 8
#define NUM_CHANNELS_CYTON 8
#define NUM_CHANNELS_CYTON_DAISY 16
#define NUM_CHANNELS_GANGLION 4
#define LED_PROG 0
#define LED_NOTIFY 5
#define GANGLION_GAIN 51
#define DEFAULT_LATENCY 10000
#define DEFAULT_MQTT_PORT 1883
// #define bit(b) (1UL << (b)) // Taken directly from Arduino.h
// Arduino JSON needs bytes for duplication
// to recalculate visit:
//   https://bblanchon.github.io/ArduinoJson/assistant/index.html
// #define ARDUINOJSON_USE_DOUBLE 1
#define ARDUINOJSON_USE_LONG_LONG 1
#define ARDUINOJSON_ADDITIONAL_BYTES_4_CHAN 115
#define ARDUINOJSON_ADDITIONAL_BYTES_8_CHAN 195
#define ARDUINOJSON_ADDITIONAL_BYTES_16_CHAN 355
#define ARDUINOJSON_ADDITIONAL_BYTES_24_CHAN 515
#define ARDUINOJSON_ADDITIONAL_BYTES_32_CHAN 675
#define MICROS_IN_SECONDS 1000000
#define SPI_MASTER_POLL_TIMEOUT_MS 100
#define SPI_TIMEOUT_CLIENT_RESPONSE 400
#define SPI_NO_MASTER 401
#define CLIENT_RESPONSE_NO_BODY_IN_POST 402
#define CLIENT_RESPONSE_MISSING_REQUIRED_CMD 403
#define NANO_VOLTS_IN_VOLTS 1000000000.0
#define MAX_JSON_BUFFER_SIZE 4000

#define ADS_GAIN_1 1
#define ADS_GAIN_2 2
#define ADS_GAIN_4 4
#define ADS_GAIN_6 6
#define ADS_GAIN_8 8
#define ADS_GAIN_12 12
#define ADS_GAIN_24 24

#define ADS_SCALE_FACTOR_VOLTS_1 0.000000536441867
#define ADS_SCALE_FACTOR_VOLTS_2 0.000000268220934
#define ADS_SCALE_FACTOR_VOLTS_4 0.000000134110467
#define ADS_SCALE_FACTOR_VOLTS_6 0.000000089406978
#define ADS_SCALE_FACTOR_VOLTS_8 0.000000067055233
#define ADS_SCALE_FACTOR_VOLTS_12 0.000000044703489
#define ADS_SCALE_FACTOR_VOLTS_24 0.000000022351744

#define MCP_SCALE_FACTOR_VOLTS 0.00000000186995

#define BOARD_TYPE_CYTON "cyton"
#define BOARD_TYPE_CYTON_DAISY "daisy"
#define BOARD_TYPE_GANGLION "ganglion"
#define BOARD_TYPE_NONE "none"

#define OUTPUT_JSON "json"
#define OUTPUT_MQTT "mqtt"
#define OUTPUT_NONE "none"
#define OUTPUT_RAW "raw"
#define OUTPUT_SERIAL "serial"
#define OUTPUT_TCP "tcp"
#define OUTPUT_WEB_SOCKETS "ws"

#define JSON_BOARD_CONNECTED "board_connected"
#define JSON_BOARD_TYPE "board_type"
#define JSON_COMMAND "command"
#define JSON_CONNECTED "connected"
#define JSON_GAINS "gains"
#define JSON_HEAP "heap"
#define JSON_LATENCY "latency"
#define JSON_MAC "mac"
#define JSON_MQTT_BROKER_ADDR "broker_address"
#define JSON_MQTT_PASSWORD "password"
#define JSON_MQTT_USERNAME "username"
#define JSON_MQTT_PORT "port"
#define JSON_NAME "name"
#define JSON_NUM_CHANNELS "num_channels"
#define JSON_REDUNDANCY "redundancy"
#define JSON_SAMPLE_NUMBERS "sample_numbers"
#define JSON_SAMPLE_NUMBER "sampleNumber"
#define JSON_TCP_DELIMITER "delimiter"
#define JSON_TCP_IP "ip"
#define JSON_TCP_OUTPUT "output"
#define JSON_TCP_PORT "port"
#define JSON_TIMESTAMPS "timestamps"
#define JSON_VERSION "version"

// Used to determine the result of processing a packet
#define PASSTHROUGH_FAIL_TOO_MANY_CHARS 0x00
#define PASSTHROUGH_FAIL_NO_CHARS 0x01
#define PASSTHROUGH_FAIL_QUEUE_FILLED 0x02
#define PASSTHROUGH_PASS 0x03
#define PROCESS_RAW_FAIL_OVERFLOW_FIRST 0x00
#define PROCESS_RAW_FAIL_OVERFLOW_FIRST_AFTER_SWITCH 0x01
#define PROCESS_RAW_FAIL_SWITCH 0x02
#define PROCESS_RAW_FAIL_OVERFLOW_MIDDLE 0x03
#define PROCESS_RAW_PASS_FIRST 0x04
#define PROCESS_RAW_PASS_SWITCH 0x05
#define PROCESS_RAW_PASS_MIDDLE 0x06
#define STREAM_PACKET_BYTE_START 0xA0
#define STREAM_PACKET_BYTE_STOP 0xC0

#define MQTT_ROUTE_KEY "openbci:eeg"

#define HTTP_ROUTE_MQTT "/mqtt"
#define HTTP_ROUTE_TCP "/tcp"
#define HTTP_ROUTE_UDP "/udp"
#define HTTP_ROUTE "/"
#define HTTP_ROUTE_CLOUD "/cloud"
#define HTTP_ROUTE_YT "/yt"
#define HTTP_ROUTE_OUTPUT_JSON "/output/json"
#define HTTP_ROUTE_OUTPUT_RAW "/output/raw"
#define HTTP_ROUTE_STREAM_START "/stream/start"
#define HTTP_ROUTE_STREAM_STOP "/stream/stop"
#define HTTP_ROUTE_VERSION "/version"
#define HTTP_ROUTE_COMMAND "/command"
#define HTTP_ROUTE_LATENCY "/latency"
#define HTTP_ROUTE_ALL "/all"
#define HTTP_ROUTE_BOARD "/board"
#define HTTP_ROUTE_WIFI "/wifi"
#define HTTP_ROUTE_WIFI_CONFIG "/wifi/config"
#define HTTP_ROUTE_WIFI_DELETE "/wifi/delete"

#define RETURN_TEXT_JSON "text/json"

#endif
